src/Storages/System/StorageSystemDetachedParts.cpp

```diff
@@ -8,108 +8,227 @@
+            if (detached_parts.empty())
+                getMoreParts();
+
+            if (detached_parts.empty())
+                return nullWhenNoRows(std::move(new_columns));
+
+            generateRows(new_columns, block_size - new_columns[0]->size());
+        }
+
+        return nullWhenNoRows(std::move(new_columns));
+    }
+
+private:
+    std::shared_ptr<SourceState> state;
+    const std::vector<UInt8> columns_mask;
+    const UInt64 block_size;
+    const bool has_bytes_on_disk_column;
+    static const size_t support_threads = 35;
+
+    StoragesInfo current_info;
+    DetachedPartsInfo detached_parts;
+
+    void getMoreParts()
+    {
+        chassert(detached_parts.empty());
+
+        current_info = state->next();
+        if (!current_info)
+            return;
+
+        detached_parts = current_info.data->getDetachedParts();
+    }
+
+    void calculatePartSizeOnDisk(size_t begin, std::vector<std::atomic<size_t>> & parts_sizes)
+    {
+        if (!has_bytes_on_disk_column)
+            return;
+
+        struct Task
+        {
+            DiskPtr disk;
+            String file;
+            std::atomic<size_t> * counter = nullptr;
+        };
+
+        struct SharedState
+        {
+            std::vector<Task> tasks;
+            std::atomic<size_t> next_task = {0};
+        };
+
+        SharedState shared_state;
+
+        for (auto p_id = begin; p_id < detached_parts.size(); ++p_id)
+        {
+            auto & p = detached_parts.at(p_id);
+            auto & disk = p.disk;
+
+            auto part_path = fs::path(MergeTreeData::DETACHED_DIR_NAME) / p.dir_name;
+            auto relative_path = fs::path(current_info.data->getRelativeDataPath()) / part_path;
+            auto listing = partFilesOnDisk(disk, relative_path);
+
+            auto * counter = &parts_sizes[p_id - begin];
+
+            for (auto & file : listing)
+                shared_state.tasks.push_back({disk, file, counter});
+        }
+
+        std::vector<std::future<void>> futures;
+        SCOPE_EXIT_SAFE({
+            /// Exceptions are not propagated
+            for (auto & future : futures)
+                if (future.valid())
+                    future.wait();
+            futures.clear();
+        });
+
+        for (size_t i = 0; i < support_threads; ++i)
+        {
+            futures.push_back(
+                scheduleFromThreadPool<void>(
+                    [&shared_state] ()
+                    {
+                        for (auto id = shared_state.next_task++; id < shared_state.tasks.size(); id = shared_state.next_task++)
+                        {
+                            auto & task = shared_state.tasks.at(id);
+                            auto size = task.disk->getFileSize(task.file);
+                            task.counter->fetch_add(size);
+                        }
+                    },
+                    IOThreadPool::get(),
+                    "DP_BytesOnDisk"));
```

Comment:
A bit about style: I think it's better to:
* call `IOThreadPool::get()` only once
* avoid writing lambda in-place in function call (even better - make a normal function)

Reply:
scheduleFromThreadPool requires r-value function without args.  shared_state has to be passed in it anyway. So it would be something like that: either std::bin(func, shared_state) or std::bin(std::mem_func(Class::func, obj), shared_state). That are equal to lambda actually.
 
IOThreadPool::get() is quite lightweight in comparison with syscalls which are made here.